dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1199 @@
+      # Get the file containers created by the previous module in the recipe
+      file_containers = self.state.GetContainers(containers.File)
+
+      # Loop over the file containers and parse the content of each to fill the
+      # resources dictionary
+      for file_container in file_containers:
+        self.logger.info(
+            f'Loading file from file system container: {file_container.path}')
+        self._ParseLogMessagesFromFileContainer(file_container)
+
+    elif self._mode == OperatingMode.ONLINE:
+      self.logger.info('Starting module in online mode.')
+      self._GetListOfResources(self._project_id)
+      self._GetResourcesMetaDataFromLogs(self._project_id)
+
+    else:
+      self.PublishMessage(
+          'Invalid operating mode. Supported modes are "online" or "offline"')
+
+    self._BuildResourcesParentRelationships()
+
+    matched_resource = self._FindResource(self._resource_name,
+                                          self._resource_type, None,
+                                          self._project_id)
+
+    if not matched_resource:
+      self.logger.error('Resource not found')
+      return
+
+    resource = matched_resource
+
+    # Save resource tree to temp file
+    output_file = tempfile.NamedTemporaryFile(mode='w',
+                                              delete=False,
+                                              encoding='utf-8',
+                                              suffix='.txt')
+    output_path = output_file.name
+
+    self.PublishMessage(f'Saving resource tree to {output_path}')
+    with open(output_path, 'w') as out_file:
+      if resource.parent:
+        out_file.write(str(resource.parent))
+      else:
+        out_file.write(str(resource))
+
+    # Dump the resource tree to CLI
+    if resource.parent:
+      self.PublishMessage(str(resource.parent))
+    else:
+      self.PublishMessage(str(resource))
+
+  def _GetListOfResources(self, project_id: str) -> None:
+    """Acquire a list of resources under a project.
+
+    Args:
+      project_id: Project id to get list of resources from.
+    """
+    # Retrieve list of disks in a project
+    self._resources_dict.update(self._RetrieveListOfDisks(project_id))
+
+    # Retrieve list of disk images in a project
+    self._resources_dict.update(self._RetrieveListOfDiskImages(project_id))
+
+    # Retrieve list of snapshots in a project
+    self._resources_dict.update(self._RetrieveListOfSnapshots(project_id))
+
+    # Retrieve list of instances in a project
+    self._resources_dict.update(self._RetrieveListOfInstances(project_id))
+
+    # Retrieve list of instance templates in a project
+    self._resources_dict.update(
+        self._RetrieveListOfInstanceTemplate(project_id))
+
+    # Retrieve list of machine images in a project
+    self._resources_dict.update(self._RetrieveListOfMachineImages(project_id))
+
+  def _GetResourcesMetaDataFromLogs(self, project_id: str) -> None:
```

Comment:
this method does a lot of things, please add high-level description to tasks it does

Reply:
Done